hudi-client/hudi-client-common/src/test/java/org/apache/hudi/table/action/TestCleanPlanner.java

```diff
@@ -0,0 +1,221 @@
+    when(mockCompletedCommitsTimeline.countInstants()).thenReturn(100);
+    when(mockHoodieTable.getCompletedCommitsTimeline()).thenReturn(mockCompletedCommitsTimeline);
+  }
+
+  @ParameterizedTest
+  @MethodSource("keepLatestByHoursArgs")
+  void keepLatestByHours(String earliestInstant, List<HoodieFileGroup> allFileGroups, List<Pair<String, Option<byte[]>>> savepoints,
+                         List<HoodieFileGroup> replacedFileGroups, Pair<Boolean, List<CleanFileInfo>> expected) {
+    // setup savepoint mocks
+    Set<String> savepointTimestamps = savepoints.stream().map(Pair::getLeft).collect(Collectors.toSet());
+    when(mockHoodieTable.getSavepointTimestamps()).thenReturn(savepointTimestamps);
+    if (!savepoints.isEmpty()) {
+      HoodieActiveTimeline activeTimeline = mock(HoodieActiveTimeline.class);
+      when(mockHoodieTable.getActiveTimeline()).thenReturn(activeTimeline);
+      for (Pair<String, Option<byte[]>> savepoint : savepoints) {
+        HoodieInstant instant = new HoodieInstant(false, HoodieTimeline.SAVEPOINT_ACTION, savepoint.getLeft());
+        when(activeTimeline.getInstantDetails(instant)).thenReturn(savepoint.getRight());
+      }
+    }
+    String partitionPath = "partition1";
+    // setup replaced file groups mocks
+    when(mockFsView.getReplacedFileGroupsBefore(earliestInstant, partitionPath)).thenReturn(replacedFileGroups.stream());
+    // setup current file groups mocks
+    when(mockFsView.getAllFileGroupsStateless(partitionPath)).thenReturn(allFileGroups.stream());
+    HoodieWriteConfig config = HoodieWriteConfig.newBuilder().withPath("/tmp")
+        .withCleanConfig(HoodieCleanConfig.newBuilder()
+            .cleanerNumHoursRetained(24)
+            .withCleanerPolicy(HoodieCleaningPolicy.KEEP_LATEST_BY_HOURS)
+            .build())
+        .build();
+    CleanPlanner<?, ?, ?, ?> cleanPlanner = new CleanPlanner<>(context, mockHoodieTable, config);
+    HoodieInstant earliestCommitToRetain = new HoodieInstant(HoodieInstant.State.COMPLETED, "COMMIT", earliestInstant);
+    Pair<Boolean, List<CleanFileInfo>> actual = cleanPlanner.getDeletePaths(partitionPath, Option.of(earliestCommitToRetain));
+    assertEquals(expected, actual);
+  }
+
+  static Stream<Arguments> keepLatestByHoursArgs() {
+    String earliestInstant = "20231204194919610";
+    // Only one file slice in the group: should still be kept even with commit earlier than "earliestInstant"
+    Arguments singleFileSliceCase = Arguments.of(
+        earliestInstant,
+        Collections.singletonList(buildFileGroup(Collections.singletonList("20231104194919610"))),
+        Collections.emptyList(),
+        Collections.emptyList(),
+        Pair.of(false, Collections.emptyList()));
+    // File group with two slices, both are before the earliestInstant. Only the latest slice should be kept.
+    HoodieFileGroup fileGroupsBeforeInstant = buildFileGroup(Arrays.asList("20231104194919610", "20231105194919610"));
```

Comment:
minor suggestion for readability of tests. 
can you declare a constant for 20231104194919610 and derive other commits times based on that? 
for eg, if earliestInstantLong variable exists, here we can generate commit times like 
earliestInstantLong + 1000, earliestInstantLong -1000. 
for ease of understanding the diff commit times in play. 

will leave it to you to take the call though.

Reply:
In order to avoid all the string/time conversions I'm just going to define them up at the top of the parameter generation as strings with names indicating their relative timing.